samples/GalleryApp/ObjectDetectorPage.xaml.cs

```diff
@@ -0,0 +1,494 @@
+                await m_frameSource.StartAsync();
+            }
+        }
+
+        /// <summary>
+        /// Print a message to the UI
+        /// </summary>
+        /// <param name="message"></param>
+        private void NotifyUser(String message)
+        {
+            if (Dispatcher.HasThreadAccess)
+            {
+                UIMessageTextBlock.Text = message;
+            }
+            else
+            {
+                Dispatcher.RunAsync(Windows.UI.Core.CoreDispatcherPriority.Normal, () => UIMessageTextBlock.Text = message).AsTask().Wait();
+            }
+        }
+
+        /// <summary>
+        /// Update media source buttons (top row)
+        /// </summary>
+        /// <param name="enableButtons"></param>
+        /// <returns></returns>
+        private async Task UpdateMediaSourceButtonsAsync(bool enableButtons)
+        {
+            if (Dispatcher.HasThreadAccess)
+            {
+                UICameraButton.IsEnabled = enableButtons;
+                UIFilePickerButton.IsEnabled = enableButtons;
+            }
+            else
+            {
+                await Dispatcher.RunAsync(Windows.UI.Core.CoreDispatcherPriority.Normal, async () => await UpdateMediaSourceButtonsAsync(enableButtons));
+            }
+        }
+
+        /// <summary>
+        /// FrameAvailable event handler
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="frame"></param>
+        private void FrameSource_FrameAvailable(object sender, VideoFrame frame)
+        {
+            // Locking behavior, so only one skill execution happens at a time
+            if (m_lock.Wait(0))
+            {
+#pragma warning disable CS4014
+                // Purposely don't await this: want handler to exit ASAP
+                // so that realtime capture doesn't wait for completion.
+                // Instead, we unlock only when processing finishes ensuring that
+                // only one execution is active at a time, dropping frames or
+                // aborting skill runs as necessary
+                Task.Run(async () =>
+                {
+                    try
+                    {
+                        await DetectObjectsAsync(frame);
+                        await DisplayFrameAndResultAsync(frame);
+                    }
+                    catch (Exception ex)
+                    {
+                        NotifyUser(ex.Message);
+                    }
+                    finally
+                    {
+                        m_lock.Release();
+                    }
+                });
+#pragma warning restore CS4014
+            }
+        }
+
+        /// <summary>
+        /// Click handler for video file button. Spawns file picker UI
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="e"></param>
+        private async void UIFilePickerButton_Click(object sender, RoutedEventArgs e)
```

Comment:
>private async void UIFilePickerButton_Click(object sender, RoutedEventArgs e) [](start = 8, length = 77)

This function looks like it can be reused between the different skills as is so I would suggest to reuse the code, rather than copy it. I am ok with submitting it this way as long as there is a refactor work is implemented afterwards.

Reply:
I will file a task for the refactor work.

---
In reply to: [297794996](https://github.com/microsoft/WindowsVisionSkillsPreview/pull/39#discussion_r297794996) [](ancestors = 297794996)